### PR TITLE
feat: add music

### PR DESCRIPTION
**Pull request by oim1**
## Description

This adds music functionality to the bot. It will play the top search result from YouTube when provided a search query or will play a link from YouTube, Spotify or Soundcloud. 

Numerous different commands - all of them are self explanatory

In order for the feature to work you must run npm install.

No API key is required. 

## Test Results

N/A for now 

## Discord Username

oim#0001

### CODE REVIEW

File: src/commands/music/queue.ts [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    executor: (msg) => {
+        const queue = distube.getQueue(msg);
+        if (!queue) {
+            return msg.channel.send('There is no queue.');
+        }
+        const queueEmbed = makeEmbed({
+            title: 'Queue',
+            description: makeLines([
+                '**Current queue:**',
+                `${queue.songs.map((song, id) => `**${id + 1}**: ${song.name} - ${song.formattedDuration}`).slice(0, 10).join('\n')}`,
```
Comment (benw202) [inline src/commands/music/queue.ts:19]: Add user who added song

**Review by benw202 — COMMENTED** [review]
Could we make the skip a vote skip - so people aren't skipping to their songs all the time

File: src/commands/music/play.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+        if (arg.length <= 1) {
+            return msg.channel.send('Please provide a song to be played');
+        }
+        if (voiceChannel) {
+            distube.play(voiceChannel, arg.join(' '), {
+                textChannel: msg.channel as TextChannel,
+                member: msg.member,
+            });
+        } else {
+            return msg.channel.send('Please join a voice channel before using this command.');
```
Comment (benw202) [inline src/commands/music/play.ts:22]: Embed please

File: src/commands/music/play.ts [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+export const play: CommandDefinition = {
+    name: ['play', 'p'],
+    description: 'Plays a song from YouTube, Spotify or Soundcloud',
+    category: CommandCategory.MUSIC,
+    executor: (msg) => {
+        const voiceChannel = msg.member?.voice?.channel;
+        const arg = msg.content.replace(/(?:\.play|\.p)\s+/, ' ').split(' ');
+        if (arg.length <= 1) {
+            return msg.channel.send('Please provide a song to be played');
```
Comment (benw202) [inline src/commands/music/play.ts:14]: Embed please

File: src/handlers/music.ts [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+            `Duration: ${song.formattedDuration}`,
+        ]),
+    });
+    queue.textChannel?.send({ embeds: [playEmbed] });
+});
+distube.on('empty', (queue) => {
+    queue.textChannel?.send('The voice channel is empty. Leaving the voice channel.');
+})
+distube.on('disconnect', (queue) => {
+    queue.textChannel?.send('Disconnecting');
```
Comment (benw202) [inline src/handlers/music.ts:39]: Embed please

File: src/handlers/music.ts [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+});
+distube.on('error', (textChannel, e) => {
+    textChannel.send(`An error has been encountered: ${e.message.slice(0, 2000)}`);
```
Comment (benw202) [inline src/handlers/music.ts:42]: Embed please

**Review by benw202 — COMMENTED** [review]
Prefer messages as embeds to make them stand out :)

**Comment by benw202** [comment]
Might be worth restricting music commands to a specific channel - or even the text channel of the VC that the bot is in

**Comment by oim1** [comment]
> Might be worth restricting music commands to a specific channel - or even the text channel of the VC that the bot is in

i plan on restricting it to the music text channel only
for testing purposes that will be added last

**Comment by ExampleWasTaken** [comment]
Given this plays the top result of a YT search query, shouldn't there be checks in place that (try to) prevent abuse? Thinking of ear rape, NSFW contents, etc.

**Comment by pdellaert** [comment]
> > Might be worth restricting music commands to a specific channel - or even the text channel of the VC that the bot is in
> 
> i plan on restricting it to the music text channel only for testing purposes that will be added last

Could you add it so we can test it? The easiest solution it to have a MUSIC entry in the `Channels` list in `constants.ts` and then refer to it. In your local environment (or mine for testing) I can just overwrite the ID with my own music channel of my test server.

**Comment by pdellaert** [comment]
Just as an fyi, running node v19, I needed to: 
```
npm uninstall @discordjs/opus
npm install @discordjs/opus@0.9.0
```

Potentially we need to add `"@discordjs/opus": "^0.9.0"` to the packages.json

Could also just be me, obviously.

**Comment by pdellaert** [comment]
Thoughts on features in order of what I think is the priority:

- Max queue size (configurable through constants): reject adding items to the queue if full (suggest max 10)
- Moderator only `.disable-music` command: 
    - entirely disables the feature until re-enabled. In case there's some abuse happening
    - Clears out the queue
    - Stops the running song immediately
    - Can be fully in memory, doesn't need to survive bot restart
- Moderator only `.enable-music` command: enables music again (doesn't restore anything)
- Moderator queue delete item: `.music-delete <number>` (or some command) . Number is the item in the queue that needs to be deleted. This could maybe even for the user to delete their own entries (and only their own). 
- Moderator forced skip command: Skip without vote
- Skip a certain number of songs with `.skip number`
- `.nowplaying` command to show the information of which song is currently playing and who requested it
- `.shuffle on/off` shuffle plays the queue randomly instead of sequential

**Comment by benw202** [comment]
Have put this back in draft for now :)

**Pull request closed without merging**